lib/poetry-dep-graph-builder.ts

```diff
@@ -0,0 +1,95 @@
+import { DepGraph, DepGraphBuilder, PkgInfo } from '@snyk/dep-graph';
+import { PoetryLockFileDependency } from './lock-file-parser';
+
+// Poetry uses the virtualenv to create an environment and this comes with these
+// packages pre-installed, therefore they won't be part of the lockfile.
+// See: https://github.com/python-poetry/poetry/issues/3075#issuecomment-703334427
+const IGNORED_DEPENDENCIES: string[] = [
+  'setuptools',
+  'distribute',
+  'pip',
+  'wheel',
+];
+
+export function build(
+  pkgDetails: PkgInfo,
+  pkgDependencyNames: string[],
+  pkgSpecs: PoetryLockFileDependency[],
+): DepGraph {
+  const builder = new DepGraphBuilder({ name: 'poetry' }, pkgDetails);
+  addDependenciesToGraph(
+    pkgDependencyNames,
+    pkgSpecs,
+    builder.rootNodeId,
+    builder,
+  );
+  return builder.build();
+}
+
+export function addDependenciesToGraph(
+  pkgNames: string[],
+  pkgSpecs: PoetryLockFileDependency[],
+  parentNodeId: string,
+  builder: DepGraphBuilder,
+) {
+  for (const pkgName of pkgNames) {
+    addDependenciesForPkg(pkgName, pkgSpecs, parentNodeId, builder);
+  }
+}
+
+export function addDependenciesForPkg(
+  pkgName: string,
+  pkgSpecs: PoetryLockFileDependency[],
+  parentNodeId: string,
+  builder: DepGraphBuilder,
+) {
+  if (IGNORED_DEPENDENCIES.indexOf(pkgName) >= 0) {
+    return;
+  }
+  // Poetry will auto-resolve dependencies with hyphens to dashes, but keep transitive reference name with underscore
+  pkgName = pkgName.replace(/_/g, '-');
+
+  const pkg = pkgLockInfoFor(pkgName, pkgSpecs);
+  if (isPkgAlreadyInGraph(pkg, builder)) {
+    builder.connectDep(parentNodeId, pkgName);
+    return;
+  }
+
+  const pkgInfo: PkgInfo = { name: pkgName, version: pkg.version };
+  builder.addPkgNode(pkgInfo, pkgName).connectDep(parentNodeId, pkgName);
+  addDependenciesToGraph(pkg.dependencies, pkgSpecs, pkgName, builder);
+}
+
+export function isPkgAlreadyInGraph(
+  pkg: PoetryLockFileDependency,
+  builder: DepGraphBuilder,
+): boolean {
+  const existingPkg = builder
+    .getPkgs()
+    .find(
+      (existingPkg) =>
+        existingPkg.name === pkg.name && existingPkg.version === pkg.version,
+    );
+  return !!existingPkg;
+}
+
+export function pkgLockInfoFor(
+  pkgName: string,
+  pkgSpecs: PoetryLockFileDependency[],
+): PoetryLockFileDependency {
+  const pkgLockInfo = pkgSpecs.find((lockItem) => {
+    return lockItem.name.toLowerCase() === pkgName.toLowerCase();
+  });
+
+  if (pkgLockInfo === undefined) {
```

Comment:
```
if (!pkgLockInfo) { ... }

Reply:
I used `undefined` to be explicit, is using `!` common enough that it reads the same way in Javascript? Not questioning, just trying to better my understanding.